flight/Modules/UAVOCrossfireTelemetry/uavocrossfiretelemetry.c

```diff
@@ -0,0 +1,235 @@
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, see <http://www.gnu.org/licenses/>
+ */
+
+#include "pios.h"
+#if defined(PIOS_INCLUDE_CROSSFIRE)
+
+#include "openpilot.h"
+#include "pios_hal.h"
+#include "pios_thread.h"
+#include "pios_modules.h"
+#include "pios_crc.h"
+#include "taskmonitor.h"
+#include "taskinfo.h"
+
+#include "uavocrossfiretelemetry.h"
+
+#include "modulesettings.h"
+#include "flightbatterystate.h"
+#include "flightbatterysettings.h"
+#include "gpsposition.h"
+#include "attitudeactual.h"
+#include "manualcontrolsettings.h"
+
+// Private constants
+#define STACK_SIZE_BYTES 600		// Reevaluate.
+#define TASK_PRIORITY				PIOS_THREAD_PRIO_LOW
+
+#define DEG2RAD(x) ((float)x * (float)M_PI / 180.0f)
+
+// Private variables
+static struct pios_thread *uavoCrossfireTelemetryTaskHandle;
+static bool module_enabled;
+
+// Crossfire receiver device
+static uintptr_t crsf_telem_dev_id;
+
+static void uavoCrossfireTelemetryTask(void *parameters);
+
+/**
+ * start the module
+ * \return -1 if start failed
+ * \return 0 on success
+ */
+static int32_t uavoCrossfireTelemetryStart(void)
+{
+	uintptr_t rcvr = PIOS_HAL_GetReceiver(MANUALCONTROLSETTINGS_CHANNELGROUPS_TBSCROSSFIRE);
+
+	// We shouldn't need to check whether it's a valid Crossfire device, because if
+	// something else was read out of the channel group, something's really borked.
+	if(rcvr) {
+		crsf_telem_dev_id = PIOS_RCVR_GetLowerDevice(rcvr);
+		if (module_enabled && (PIOS_Crossfire_InitTelemetry(crsf_telem_dev_id) == 0)) {
+			// Start task
+			uavoCrossfireTelemetryTaskHandle = PIOS_Thread_Create(
+					uavoCrossfireTelemetryTask, "uavoCrossfireTelemetry",
+					STACK_SIZE_BYTES, NULL, TASK_PRIORITY);
+			TaskMonitorAdd(TASKINFO_RUNNING_UAVOCROSSFIRETELEMETRY,
+					uavoCrossfireTelemetryTaskHandle);
+			return 0;
+		}
+	}
+
+	return -1;
+}
+
+static int32_t uavoCrossfireTelemetryInitialize(void)
+{
+	module_enabled = PIOS_Modules_IsEnabled(PIOS_MODULE_UAVOCROSSFIRETELEMETRY); 
+	return 0;
+}
+MODULE_INITCALL(uavoCrossfireTelemetryInitialize, uavoCrossfireTelemetryStart)
+
+// Stupid big endian stuff.
+#define WRITE_VAL(buf,p,x)				{ typeof(x) v = x; uint8_t *q = (uint8_t*)&v, n = sizeof(v); do{ buf[p++] = q[--n]; } while(n); }
+#define WRITE_VAL16(buf,p,x)			{ typeof(x) v = x; uint8_t *q = (uint8_t*)&v; buf[p++] = q[1]; buf[p++] = q[0]; }
+#define WRITE_VAL32(buf,p,x)			{ typeof(x) v = x; uint8_t *q = (uint8_t*)&v; buf[p++] = q[3]; buf[p++] = q[2]; buf[p++] = q[1]; buf[p++] = q[0]; }
+
+static int crsftelem_create_attitude(uint8_t *buf)
+{
+	int ptr = 0;
```

Comment:
not a huge problem / doesn't need to be fixed, but I'd prefer `pos` to `ptr` since it is not, in fact, a pointer.